File: src/main/java/club/sk1er/patcher/util/name/NameFetcher.java
package club.sk1er.patcher.util.name;

import club.sk1er.patcher.Patcher;
import gg.essential.api.EssentialAPI;
import gg.essential.api.utils.Multithreading;
import gg.essential.api.utils.mojang.Name;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public class NameFetcher {
    private final List<String> names = new ArrayList<>();
    private final DateFormat format = new SimpleDateFormat("MM/dd/yyyy");
    private UUID uuid = null;
    private String name = null;

    public void execute(String username) {
        execute(username, true);
    }

    public void execute(String username, boolean async) {
        try {
            if (username.isEmpty()) {
                return;
            }

            Runnable fetchNames = () -> {
                name = username;
                uuid = null;
                try {
                    CompletableFuture<UUID> uuid = EssentialAPI.getMojangAPI().getUUID(username);
                    if (uuid == null) return;
                    this.uuid = uuid.get();
                } catch (Exception e) {
                    Patcher.instance.getLogger().warn("Failed fetching UUID.", e);
                    return;
                }

                names.clear();
                if (uuid != null) {
                    List<Name> nameHistory = EssentialAPI.getMojangAPI().getNameHistory(uuid);
                    if (nameHistory == null || nameHistory.isEmpty()) return;
                    name = nameHistory.get(nameHistory.size() - 1).getName();

                    for (final Name history : nameHistory) {
                        final String name = history.getName();
                        //noinspection ConstantConditions
                        if (history.getChangedToAt() == 0) {
                            names.add(name);
                        } else {
                            names.add(String.format("%s » %s", name, format.format(history.getChangedToAt())));
                        }
                    }
                } else {
                    names.add("Failed to fetch " + username + "'s names");
                }
            };

            if (async) {
                Multithreading.runAsync(fetchNames);
            } else {
                fetchNames.run();
            }
        } catch (Exception e) {
            Patcher.instance.getLogger().warn("User catch failed, tried fetching {}.", username, e);
        }
    }

    public List<String> getNames() {
        return names;
    }

    public UUID getUuid() {
        return uuid;
    }

    public String getName() {
        return name;
    }
}

File: src/main/kotlin/club/sk1er/patcher/util/NameFetcher.kt
package club.sk1er.patcher.util

import club.sk1er.patcher.Patcher
import club.sk1er.patcher.coroutines.MCDispatchers
import gg.essential.api.EssentialAPI
import gg.essential.api.utils.mojang.Name
import gg.essential.elementa.utils.ObservableList
import kotlinx.coroutines.launch
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class NameFetcher {
    val names = ObservableList(mutableListOf<Name>())
    private val format: DateFormat = SimpleDateFormat("MM/dd/yyyy")
    var uuid: UUID? = null
        private set
    var currentName: String? = null
        private set
    var callback: () -> Unit = {}

    fun execute(username: String, async: Boolean = true) {
        try {
            if (username.isEmpty()) {
                return
            }
            val fetchNames = Runnable {
                currentName = username
                uuid = null
                names.clear()
                try {
                    val uuid = EssentialAPI.getMojangAPI().getUUID(username) ?: return@Runnable
                    this.uuid = uuid.get()
                } catch (e: Exception) {
                    Patcher.instance.logger.warn("Failed fetching UUID.", e)
                    return@Runnable
                }
                if (uuid != null) {
                    val nameHistory =
                        EssentialAPI.getMojangAPI().getNameHistory(uuid)
                    if (nameHistory == null || nameHistory.isEmpty()) return@Runnable
                    currentName = nameHistory[nameHistory.size - 1]!!.name
                    nameHistory.forEach {
                        if (it == null) return@forEach
                        names.add(it)
                    }
                } else {
                    names.add(Name("Failed to fetch $username's names", 0))
                }
            }
            if (async) {
                MCDispatchers.PATCHER_SCOPE.launch {
                    fetchNames.run()
                }.invokeOnCompletion {
                    callback()
                }
            } else {
                fetchNames.run()
                callback()
            }
        } catch (e: Exception) {
            Patcher.instance.logger.warn("User catch failed, tried fetching {}.", username, e)
        }
    }

    fun getDate(index: Int): String {
        return format.format(names[index].changedToAt?.let { Date(it) })
    }
}